extends Node2D


class_name MessageWindowProgram


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# ãƒ¡ãƒ¢ãƒªé–¢é€£
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


# çŠ¶æ…‹æ©Ÿæ¢°
var statemachine_of_message_window = load("res://ğŸ‹scripts/ğŸª‘grayscale_ğŸ‰visual_novel/ğŸ±message_windows/statemachines/message_window.gd").new()

# ãƒšãƒ¼ã‚¸é€ã‚Šæ™‚ã®å¯¾å¿œé–¢æ•°
var on_message_window_page_forward = func():
	pass

# å¯è¦–æ€§
var is_appear = true

# ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ãƒ»ã‚¹ãƒ”ãƒ¼ãƒ‰ï¼ˆï¼‘ç§’å½“ãŸã‚Šã®æ–‡å­—æ•°ï¼‰
var msg_speed = 20.0

# ãƒ†ã‚­ã‚¹ãƒˆãƒ–ãƒ­ãƒƒã‚¯æ±ç”¨
#	ã‚»ã‚¯ã‚·ãƒ§ãƒ³ã®ä½•ç•ªç›®ï¼ˆï¼é–‹å§‹ï¼‰ã‚’æ¬¡ã«èª­è¾¼ã‚€ã‹
var section_item_index = 0
#	è¡¨ç¤ºã—ãŸã„æ–‡ç« 
var text_block_buffer = ""
#	ã‚¿ã‚¤ãƒ—ãƒ©ã‚¤ã‚¿ãƒ¼ã®ã‚«ã‚¦ãƒ³ãƒˆ
var count_of_typewriter = 0

# é¸æŠè‚¢
#	é¸æŠè‚¢ã®è¡Œç•ªå·ç”¨é…åˆ—ã€‚ç„¡ã‘ã‚Œã°ãƒŒãƒ«
var choices_row_numbers = null
#	é¸æŠè‚¢é…åˆ—ã®ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹
var choices_index = 0


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# è¦ªãƒ‘ã‚¹é–¢é€£
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


# ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ãƒ»ã‚¦ã‚£ãƒ³ãƒ‰ã‚¦ã®è»¸å–å¾—
func hub():
	pass


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# æº–å‚™
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


func _ready():
	
	# çŠ¶æ…‹æ©Ÿæ¢°ã®ã‚»ãƒƒãƒˆã‚¢ãƒƒãƒ—
	self.statemachine_of_message_window.on_talked_2 = self.on_talked_2
	self.statemachine_of_message_window.on_page_forward = self.on_page_forward
	self.statemachine_of_message_window.on_all_characters_pushed = self.on_all_characters_pushed
	self.statemachine_of_message_window.on_all_pages_flushed = self.on_all_pages_flushed


# çŠ¶æ…‹é·ç§»ã™ã‚‹ã ã‘
func on_talked_2():

	var illust_a = self.hub().get_programs_hub().get_illust(self.name)

	# è¡¨ç¤º
	self.set_visible_subtree(true)
	illust_a.modulate.a = 1.0	# ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸è¿½åŠ ã«ã‚ˆã‚‹ä¸é€æ˜åŒ–

	var message_window_gui = self.hub().get_programs_hub().get_current_message_window_gui()

	# é¸æŠè‚¢ãªã‚‰
	if message_window_gui.is_choices():
		print("ï¼»ä¼è¨€çª“ã€€â€" + self.name + "â€ï¼½ã€€é¸æŠè‚¢é–‹å§‹")
		# ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã‚¨ãƒ³ãƒ‰ãƒ»ãƒ–ãƒªãƒ³ã‚«ãƒ¼ã€€çŠ¶æ…‹æ©Ÿæ¢°ï¼»æ±ºã‚ãŸï¼½
		self.hub().get_blinker_triangle(self.name).statemachine_of_end_of_message_blinker.decide()
		self.hub().get_blinker_underscore(self.name).statemachine_of_end_of_message_blinker.decide()
		
		# ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã‚¨ãƒ³ãƒ‰ãƒ»ãƒ–ãƒªãƒ³ã‚«ãƒ¼ã€€çŠ¶æ…‹æ©Ÿæ¢°ï¼»è€ƒãˆã‚‹ï¼½
		self.hub().get_choices_cursor(self.name).statemachine_of_end_of_message_blinker.think()
	
	else:
		print("ï¼»ä¼è¨€çª“ã€€â€" + self.name + "â€ï¼½ã€€å°è©é–‹å§‹")
		# ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã‚¨ãƒ³ãƒ‰ãƒ»ãƒ–ãƒªãƒ³ã‚«ãƒ¼ã€€çŠ¶æ…‹æ©Ÿæ¢°ï¼»æ±ºã‚ãŸï¼½
		self.hub().get_choices_cursor(self.name).statemachine_of_end_of_message_blinker.decide()
		
		# ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã‚¨ãƒ³ãƒ‰ãƒ»ãƒ–ãƒªãƒ³ã‚«ãƒ¼ã€€çŠ¶æ…‹æ©Ÿæ¢°ï¼»è€ƒãˆã‚‹ï¼½
		self.hub().get_blinker_triangle(self.name).statemachine_of_end_of_message_blinker.think()
		self.hub().get_blinker_underscore(self.name).statemachine_of_end_of_message_blinker.think()


# ãƒšãƒ¼ã‚¸é€ã‚Š
# çŠ¶æ…‹é·ç§»æ©Ÿæ¢°ã‹ã‚‰å‘¼å‡ºã•ã‚Œã‚‹
func on_page_forward():
	var message_window_gui = self.hub().get_programs_hub().get_current_message_window_gui()

	# é¸æŠè‚¢ãƒ¢ãƒ¼ãƒ‰ãªã‚‰
	if message_window_gui.is_choices():

		# ã‚«ãƒ¼ã‚½ãƒ«éŸ³
		self.hub().get_programs_hub().get_instruction(&"ğŸ“—Se").play_se("ğŸ””é¸æŠè‚¢ç¢ºå®šéŸ³")

		var row_number = message_window_gui.get_row_number_of_choices()
		print("ï¼»ä¼è¨€çª“ã€€â€" + self.name + "â€ï¼½ã€€é¸ã‚“ã é¸æŠè‚¢è¡Œç•ªå·ï¼šï¼»" + str(row_number) + "ï¼½")

		# é¸æŠè‚¢ã®è¡Œç•ªå·ã‚’ã€ä¸Šä½ãƒãƒ¼ãƒ‰ã¸ã‚¨ã‚¹ã‚«ãƒ¬ãƒ¼ã‚·ãƒ§ãƒ³ã—ã¾ã™
		self.hub().get_programs_hub().on_choice_selected(row_number)

		# é¸æŠè‚¢ã¯ãŠå½¹å¾¡å…
		message_window_gui.choices_row_numbers = null
		
	else:
		print("ï¼»ä¼è¨€çª“ã€€â€" + self.name + "â€ï¼½ã€€ãƒšãƒ¼ã‚¸é€ã‚Š")

		# åŠ¹æœéŸ³
		self.hub().get_programs_hub().get_instruction(&"ğŸ“—Se").play_se("ğŸ””ãƒšãƒ¼ã‚¸ã‚ãã‚ŠéŸ³")
		
		# ãƒšãƒ¼ã‚¸é€ã‚Šã‚’ã—ãŸã“ã¨ã‚’ã€å‘¼å‡ºã—å…ƒã¸ä¼ãˆã‚‹
		self.on_message_window_page_forward.call()

	# ç©ºã£ã½ã®ã‚¦ã‚£ãƒ³ãƒ‰ã‚¦ã‚’æ®‹ã—ã¦ã€æ¬¡ã®æŒ‡ç¤ºã‚’å¾…ã¡ã¾ã™
	# ãƒ†ã‚­ã‚¹ãƒˆãƒ–ãƒ­ãƒƒã‚¯
	var text_block_node = self.hub().get_text_block(self.name)
	if true:
		# ãƒ†ã‚­ã‚¹ãƒˆãŒç©ºã£ã½
		text_block_node.text = ""
		# å…¨ã¦ã®ãƒ–ãƒªãƒ³ã‚«ãƒ¼ã€€çŠ¶æ…‹æ©Ÿæ¢°ï¼»æ±ºã‚ãŸï¼½
		self.hub().get_blinker_triangle(self.name).statemachine_of_end_of_message_blinker.decide()
		self.hub().get_blinker_underscore(self.name).statemachine_of_end_of_message_blinker.decide()
		self.hub().get_choices_cursor(self.name).statemachine_of_end_of_message_blinker.decide()


func on_all_characters_pushed():
	var message_window_gui = self.hub().get_programs_hub().get_current_message_window_gui()

	# é¸æŠè‚¢
	if message_window_gui.is_choices():
		# æ–‡æœ«ãƒ–ãƒªãƒ³ã‚«ãƒ¼	çŠ¶æ…‹æ©Ÿæ¢°ï¼»è€ƒãˆã‚‹ï¼½
		self.hub().get_choices_cursor(self.name).statemachine_of_end_of_message_blinker.think()

	# ãã‚Œä»¥å¤–
	else:
		# æ–‡æœ«ãƒ–ãƒªãƒ³ã‚«ãƒ¼	çŠ¶æ…‹æ©Ÿæ¢°ï¼»è€ƒãˆã‚‹ï¼½
		self.hub().get_blinker_triangle(self.name).statemachine_of_end_of_message_blinker.think()
		self.hub().get_blinker_underscore(self.name).statemachine_of_end_of_message_blinker.think()


# åˆæœŸåŒ–
#	ã‚¦ã‚£ãƒ³ãƒ‰ã‚¦ãŒå­˜åœ¨ã—ãªã„çŠ¶æ…‹ã«æˆ»ã—ã¾ã™
func on_all_pages_flushed():
	var message_window_gui = self.hub().get_programs_hub().get_current_message_window_gui()

	print("ï¼»ä¼è¨€çª“ã€€â€" + self.name + "â€ï¼½ã€€ã‚ªãƒ³ãƒ»ã‚ªãƒ¼ãƒ«ãƒ»ãƒšãƒ¼ã‚¸ã‚ºãƒ»ãƒ•ã‚£ãƒ‹ãƒƒã‚·ãƒ¥ãƒ‰ï¼½ï¼ˆéè¡¨ç¤ºï¼‰")
	var illust_a = self.hub().get_programs_hub().get_illust(self.name)

	# ãƒ†ã‚­ã‚¹ãƒˆãƒ–ãƒ­ãƒƒã‚¯
	var text_block_node = self.hub().get_text_block(self.name)
	# ãƒ†ã‚­ã‚¹ãƒˆãŒç©ºã£ã½
	text_block_node.text = ""

	# é¸æŠè‚¢
	if message_window_gui.is_choices():
		# å…¨ã¦ã®ãƒ–ãƒªãƒ³ã‚«ãƒ¼ã€€çŠ¶æ…‹æ©Ÿæ¢°ï¼»æ±ºã‚ãŸï¼½
		self.hub().get_choices_cursor(self.name).statemachine_of_end_of_message_blinker.decide()
	else:
		# å…¨ã¦ã®ãƒ–ãƒªãƒ³ã‚«ãƒ¼ã€€çŠ¶æ…‹æ©Ÿæ¢°ï¼»æ±ºã‚ãŸï¼½
		self.hub().get_blinker_triangle(self.name).statemachine_of_end_of_message_blinker.decide()
		self.hub().get_blinker_underscore(self.name).statemachine_of_end_of_message_blinker.decide()

	# ã“ã®è¦ç´ ã®åˆæœŸçŠ¶æ…‹ã¯ã€éè¡¨ç¤ºã€é€æ˜
	self.set_visible_subtree(false)
	illust_a.modulate.a = 0.0	# åˆæœŸåŒ–ã«ã‚ˆã‚‹é€æ˜åŒ–


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# ãã®ä»–
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


# ã‚µãƒ–ãƒ„ãƒªãƒ¼ã® visible ã‚’è¨­å®š
func set_visible_subtree(
		visible_flag):			# bool

	var illust_a = self.hub().get_programs_hub().get_illust(self.name)
	print("ï¼»ä¼è¨€çª“ã€€â€" + self.name + "â€ï¼½ã€€ç¾å¯è¦–æ€§ï¼š" + str(illust_a.visible) + "ã€€æ¬¡å¯è¦–æ€§ï¼š" + str(visible_flag))

	# è¦‹ã›ã‚ï¼ˆtrueï¼‰ ã¨ã„ã†æŒ‡ç¤ºã®ã¨ãã€è¦‹ãˆã¦ã‚Œã°ï¼ˆtrueï¼‰ ã€ä½•ã‚‚ã—ãªã„ï¼ˆpassï¼‰ã€‚
	# éš ã›ã€€ï¼ˆfalseï¼‰ã¨ã„ã†æŒ‡ç¤ºã®ã¨ãã€è¦‹ãˆã¦ã‚Œã°ï¼ˆtrueï¼‰ ã€éš ã™ã€€ã€€ã€€ï¼ˆfalseï¼‰ã€‚
	# è¦‹ã›ã‚ï¼ˆtrueï¼‰ ã¨ã„ã†æŒ‡ç¤ºã®ã¨ãã€éš ã‚Œã¦ã‚Œã°ï¼ˆfalseï¼‰ã€è¦‹ã›ã‚‹ã€€ã€€ï¼ˆtrueï¼‰ã€‚
	# éš ã›ã€€ï¼ˆfalseï¼‰ã¨ã„ã†æŒ‡ç¤ºã®ã¨ãã€éš ã‚Œã¦ã‚Œã°ï¼ˆfalseï¼‰ã€ä½•ã‚‚ã—ãªã„ï¼ˆpassï¼‰
	if visible_flag != illust_a.visible:

		print("ï¼»ä¼è¨€çª“ã€€â€" + self.name + "â€ï¼½ã€€å¯è¦–æ€§ï¼š" + str(visible_flag))

		illust_a.visible = visible_flag
		self.hub().get_canvas_layer(self.name).visible = visible_flag

		# å­ãƒãƒ¼ãƒ‰
		for child in self.hub().get_text_block(self.name).get_children():
			if child.has_method("set_visible_subtree"):
				child.set_visible_subtree(visible_flag)
